File: Runtime/TransitionSlopeMultiplierMap.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.Core
Imports SyncroSim.StochasticTime

Class TransitionSlopeMultiplierMap
    Inherits STSimMapBase3(Of SortedList(Of Integer, TransitionSlopeMultiplier))

    Private m_DistributionProvider As STSimDistributionProvider

    Public Sub New(
        ByVal scenario As Scenario,
        ByVal multipliers As TransitionSlopeMultiplierCollection,
        ByVal distributionProvider As STSimDistributionProvider)

        MyBase.New(scenario)

        Me.m_DistributionProvider = distributionProvider

        For Each Item As TransitionSlopeMultiplier In multipliers
            Me.AddSlopeMultiplier(Item)
        Next

    End Sub

    Public Function GetSlopeMultiplier(
        ByVal transitionGroupId As Integer,
        ByVal stratumId As Integer,
        ByVal secondaryStratumId As Nullable(Of Integer),
        ByVal iteration As Integer,
        ByVal timestep As Integer,
        ByVal slope As Double) As Double

        Dim lst As SortedList(Of Integer, TransitionSlopeMultiplier) =
            Me.GetItem(
                transitionGroupId,
                stratumId,
                secondaryStratumId,
                iteration,
                timestep)

        If (lst Is Nothing) Then
            Return 1.0
        End If

        If (lst.Count = 1) Then

            Dim tsm As TransitionSlopeMultiplier = lst.First.Value
            tsm.Sample(iteration, timestep, Me.m_DistributionProvider, DistributionFrequency.Always)

            Return tsm.CurrentValue.Value

        End If

        Dim SlopeInt As Integer = CInt(slope)

        If (lst.ContainsKey(SlopeInt)) Then

            Dim tsm As TransitionSlopeMultiplier = lst(SlopeInt)
            tsm.Sample(iteration, timestep, Me.m_DistributionProvider, DistributionFrequency.Always)

            Return tsm.CurrentValue.Value

        End If

        Dim PrevKey As Integer = -91
        Dim ThisKey As Integer = -91

        For Each k As Integer In lst.Keys

            Debug.Assert(k <> SlopeInt)

            If (k > SlopeInt) Then

                ThisKey = k
                Exit For

            End If

            PrevKey = k

        Next

        If (PrevKey = -91) Then

            Dim tsm As TransitionSlopeMultiplier = lst.First.Value
            tsm.Sample(iteration, timestep, Me.m_DistributionProvider, DistributionFrequency.Always)

            Return tsm.CurrentValue.Value

        End If

        If (ThisKey = -91) Then

            Dim tsm As TransitionSlopeMultiplier = lst.Last.Value
            tsm.Sample(iteration, timestep, Me.m_DistributionProvider, DistributionFrequency.Always)

            Return tsm.CurrentValue.Value

        End If

        Dim PrevMult As TransitionSlopeMultiplier = lst(PrevKey)
        Dim ThisMult As TransitionSlopeMultiplier = lst(ThisKey)

        PrevMult.Sample(iteration, timestep, Me.m_DistributionProvider, DistributionFrequency.Always)
        ThisMult.Sample(iteration, timestep, Me.m_DistributionProvider, DistributionFrequency.Always)

        Return Interpolate(
            PrevKey,
            PrevMult.CurrentValue.Value,
            ThisKey,
            ThisMult.CurrentValue.Value,
            slope)

    End Function

    Private Sub AddSlopeMultiplier(ByVal item As TransitionSlopeMultiplier)

        Dim l As SortedList(Of Integer, TransitionSlopeMultiplier) =
            Me.GetItemExact(
                item.TransitionGroupId,
                item.StratumId,
                item.SecondaryStratumId,
                item.Iteration,
                item.Timestep)

        If (l Is Nothing) Then

            l = New SortedList(Of Integer, TransitionSlopeMultiplier)

            Me.AddItem(
                item.TransitionGroupId,
                item.StratumId,
                item.SecondaryStratumId,
                item.Iteration,
                item.Timestep,
                l)

        End If

        l.Add(CInt(item.Slope), item)
        Debug.Assert(Me.HasItems())

    End Sub

End Class

File: src/Runtime/TransitionSlopeMultiplierMap.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2023 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System;
using System.Linq;
using SyncroSim.Core;
using SyncroSim.StochasticTime;
using System.Collections.Generic;
using System.Diagnostics;

namespace SyncroSim.STSim
{
    internal class TransitionSlopeMultiplierMap : STSimMapBase4<SortedList<int, TransitionSlopeMultiplier>>
    {
        private STSimDistributionProvider m_DistributionProvider;

        public TransitionSlopeMultiplierMap(
            Scenario scenario, 
            TransitionSlopeMultiplierCollection multipliers,
            STSimDistributionProvider distributionProvider) : base(scenario)
        {
            this.m_DistributionProvider = distributionProvider;

            foreach (TransitionSlopeMultiplier Item in multipliers)
            {
                this.AddSlopeMultiplier(Item);
            }
        }

        public double GetSlopeMultiplier(
            int transitionGroupId, int stratumId, int? secondaryStratumId, int? tertiaryStratumId, 
            int iteration, int timestep, double slope)
        {
            SortedList<int, TransitionSlopeMultiplier> lst = this.GetItem(
                transitionGroupId, stratumId, secondaryStratumId, tertiaryStratumId, iteration, timestep);

            if (lst == null)
            {
                return 1.0;
            }

            if (lst.Count == 1)
            {
                TransitionSlopeMultiplier tsm = lst.First().Value;
                tsm.Sample(iteration, timestep, this.m_DistributionProvider, DistributionFrequency.Always);

                return tsm.CurrentValue.Value;
            }

            int SlopeInt = Convert.ToInt32(slope);

            if (lst.ContainsKey(SlopeInt))
            {
                TransitionSlopeMultiplier tsm = lst[SlopeInt];
                tsm.Sample(iteration, timestep, this.m_DistributionProvider, DistributionFrequency.Always);

                return tsm.CurrentValue.Value;
            }

            int PrevKey = -91;
            int ThisKey = -91;

            foreach (int k in lst.Keys)
            {
                Debug.Assert(k != SlopeInt);

                if (k > SlopeInt)
                {
                    ThisKey = k;
                    break;
                }

                PrevKey = k;
            }

            if (PrevKey == -91)
            {
                TransitionSlopeMultiplier tsm = lst.First().Value;
                tsm.Sample(iteration, timestep, this.m_DistributionProvider, DistributionFrequency.Always);

                return tsm.CurrentValue.Value;
            }

            if (ThisKey == -91)
            {
                TransitionSlopeMultiplier tsm = lst.Last().Value;
                tsm.Sample(iteration, timestep, this.m_DistributionProvider, DistributionFrequency.Always);

                return tsm.CurrentValue.Value;
            }

            TransitionSlopeMultiplier PrevMult = lst[PrevKey];
            TransitionSlopeMultiplier ThisMult = lst[ThisKey];

            PrevMult.Sample(iteration, timestep, this.m_DistributionProvider, DistributionFrequency.Always);
            ThisMult.Sample(iteration, timestep, this.m_DistributionProvider, DistributionFrequency.Always);

            return Statistics.Interpolate(PrevKey, PrevMult.CurrentValue.Value, ThisKey, ThisMult.CurrentValue.Value, slope);
        }

        private void AddSlopeMultiplier(TransitionSlopeMultiplier item)
        {
            SortedList<int, TransitionSlopeMultiplier> l = this.GetItemExact(item.TransitionGroupId, item.StratumId, item.SecondaryStratumId, item.TertiaryStratumId, item.Iteration, item.Timestep);

            if (l == null)
            {
                l = new SortedList<int, TransitionSlopeMultiplier>();

                this.AddItem(item.TransitionGroupId, item.StratumId, item.SecondaryStratumId, item.TertiaryStratumId, item.Iteration, item.Timestep, l);
            }

            l.Add(Convert.ToInt32(item.Slope), item);
            Debug.Assert(this.HasItems);
        }
    }
}
